core/templates/pages/skill-editor-page/services/question-creation.service.ts

```diff
@@ -0,0 +1,266 @@
+      skillIdToRubricsObject[skillId] = skill.getRubrics();
+
+      newQuestionSkillIds = [skillId];
+      var currentMode = MODE_SELECT_DIFFICULTY;
+
+      var linkedSkillsWithDifficulty = [];
+      newQuestionSkillIds.forEach(function(skillId) {
+        linkedSkillsWithDifficulty.push(
+          SkillDifficultyObjectFactory.create(
+            skillId, '', DEFAULT_SKILL_DIFFICULTY));
+      });
+      groupedSkillSummaries = (
+        SkillEditorStateService.getGroupedSkillSummaries());
+      var sortedSkillSummaries =
+          groupedSkillSummaries.current.concat(
+            groupedSkillSummaries.others);
+      var countOfSkillsToPrioritize =
+          groupedSkillSummaries.current.length;
+      var allSkillSummaries = sortedSkillSummaries.map(
+        function(summary) {
+          summary.isSelected = false;
+          return summary;
+        });
+      $uibModal.open({
+        templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+          '/pages/topic-editor-page/modal-templates/' +
+            'select-skill-and-difficulty-modal.template.html'),
+        backdrop: true,
+        resolve: {
+          allSkillSummaries: () => allSkillSummaries,
+          countOfSkillsToPrioritize: () => countOfSkillsToPrioritize,
+          currentMode: () => currentMode,
+          linkedSkillsWithDifficulty: () => linkedSkillsWithDifficulty,
+          skillIdToRubricsObject: () => skillIdToRubricsObject
+        },
+        controller: 'QuestionsListSelectSkillAndDifficultyModalController'
+      }).result.then(function(linkedSkillsWithDifficulty) {
+        newQuestionSkillIds = [];
+        newQuestionSkillDifficulties = [];
+        linkedSkillsWithDifficulty.forEach(
+          function(linkedSkillWithDifficulty) {
+            newQuestionSkillIds.push(
+              linkedSkillWithDifficulty.getId());
+            newQuestionSkillDifficulties.push(
+              linkedSkillWithDifficulty.getDifficulty());
+          });
+
+        populateMisconceptions();
+        if (AlertsService.warnings.length === 0) {
+          initializeNewQuestionCreation();
+        }
+      }, function() {
+        // Note to developers:
+        // This callback is triggered when the Cancel button is clicked.
+        // No further action is needed.
+      });
+    };
+
+    var initializeNewQuestionCreation = function() {
+      question =
+          QuestionObjectFactory.createDefaultQuestion(newQuestionSkillIds);
+      questionId = question.getId();
+      questionStateData = question.getStateData();
+      questionIsBeingUpdated = false;
+      newQuestionIsBeingCreated = true;
+      openQuestionEditor(newQuestionSkillDifficulties[0]);
+    };
+
+    var getDifficultyString = function(difficulty) {
+      if (difficulty === 0.3) {
+        return SKILL_DIFFICULTIES[0];
+      } else if (difficulty === 0.6) {
+        return SKILL_DIFFICULTIES[1];
+      } else {
+        return SKILL_DIFFICULTIES[2];
+      }
+    };
```

Comment:
Looks like this logic is used in multiple places.. eg. questions-list.directive, question-suggestion-review-modal.controller, and question-suggestion-review-modal.controller. Can we create a common method and reuse it in these places?

Reply:
Refractored.

Reply:
Can you point me to where this is done?